Wikipedia/Code/ArticleViewController+LinkPreviewing.swift

```diff
@@ -0,0 +1,234 @@
+            shareActivityViewController.excludedActivityTypes = [.addToReadingList]
+            vc.articlePreviewingDelegate?.shareArticlePreviewActionSelected(with: vc, shareActivityController: shareActivityViewController)
+        }
+        
+        var actions = [readAction, saveAction]
+        
+        if article.location != nil {
+            let placeActionTitle = WMFLocalizedString("page-location", value: "View on a map", comment: "Label for button used to show an article on the map")
+            let placeAction = UIPreviewAction(title: placeActionTitle, style: .default) { (action, vc) in
+                guard let vc = vc as? ArticleViewController else {
+                    return
+                }
+                vc.articlePreviewingDelegate?.viewOnMapArticlePreviewActionSelected(with: vc)
+            }
+            actions.append(placeAction)
+        }
+        
+        actions.append(shareAction)
+        
+        return actions
+    }
+    
+    @available(iOS 13.0, *)
+    func previewMenuElements(for previewViewController: UIViewController, suggestedActions: [UIMenuElement]) -> UIMenu? {
+        guard let vc = previewViewController as? ArticleViewController else {
+            return nil
+        }
+        let legacyActions = vc.previewActions
+        let menuItems = legacyActions.map { (legacyAction) -> UIMenuElement in
+            return UIAction(title: legacyAction.title) { (action) in
+                legacyAction.handler(legacyAction, previewViewController)
+            }
+        }
+        return UIMenu(title: "", image: nil, identifier: nil, options: [], children: menuItems)
+    }
+    
+    func commitPreview(of viewControllerToCommit: UIViewController) {
+        if let vc = viewControllerToCommit as? ArticleViewController {
+            vc.wmf_removePeekableChildViewControllers()
+            push(vc, animated: true)
+        } else {
+            if let vc = viewControllerToCommit as? WMFImageGalleryViewController {
+                vc.setOverlayViewTopBarHidden(false)
+            }
+            presentEmbedded(viewControllerToCommit, style: .gallery)
+        }
+    }
+}
+
+extension ArticleViewController: WKUIDelegate {
+    // MARK: Context menus
+    
+    @available(iOS 13.0, *)
+    func webView(_ webView: WKWebView, contextMenuConfigurationForElement elementInfo: WKContextMenuElementInfo, completionHandler: @escaping (UIContextMenuConfiguration?) -> Void) {
+        let nullConfig = UIContextMenuConfiguration(identifier: nil, previewProvider: { () -> UIViewController? in
+            return nil
+        }) { (_) -> UIMenu? in
+            return nil
+        }
+        
+        let bail = {
+            completionHandler(nullConfig)
+        }
+        
+        guard let linkURL = elementInfo.linkURL else {
+            bail()
+            return
+        }
+        
+        getPeekViewControllerAsync(for: linkURL) { (peekParentVC) in
```

Comment:
@joewalsh since this is async, do you think we also need a timeout here?

Reply:
Good catch, fixed!